chains_test.go

```diff
@@ -0,0 +1,430 @@
+
+		req.NoError(err)
+		req.NotNil(ret)
+		req.Len(ret, 1)
+		req.Len(ret[0], 1)
+		req.Equal(ret[0][0], leaf.cert)
+	})
+
+	t.Run("returns 1 chain of 1 cert with 1 root given only 1 leaf all without AKIDs", func(t *testing.T) {
+		req := require.New(t)
+		root := newRootCa()
+		leaf := root.NewLeafWithoutAKID()
+
+		ret, err := AssembleServerChains([]*x509.Certificate{leaf.cert})
+
+		req.NoError(err)
+		req.NotNil(ret)
+		req.Len(ret, 1)
+		req.Len(ret[0], 1)
+		req.Equal(ret[0][0], leaf.cert)
+	})
+
+	t.Run("returns 2 chains for two different CAs all with AKIDs", func(t *testing.T) {
+		req := require.New(t)
+
+		root1 := newRootCa()
+		leaf1 := root1.NewLeafWithAKID()
+
+		root2 := newRootCa()
+		intermediate2 := root2.NewIntermediateWithAKID()
+		leaf2 := intermediate2.NewLeafWithAKID()
+
+		ret, err := AssembleServerChains([]*x509.Certificate{root1.cert, leaf1.cert, root2.cert, intermediate2.cert, leaf2.cert})
+
+		req.NoError(err)
+		req.NotNil(ret)
+		req.Len(ret, 2)
+
+		// root1 -> leaf2
+		req.Len(ret[0], 2)
+		req.Equal(ret[0][0], leaf1.cert)
+		req.Equal(ret[0][1], root1.cert)
+
+		// root2 -> intermediate2 -> leaf 2
+
+		req.Len(ret[1], 3)
+		req.Equal(ret[1][0], leaf2.cert)
+		req.Equal(ret[1][1], intermediate2.cert)
+		req.Equal(ret[1][2], root2.cert)
+	})
+
+	t.Run("returns 2 chains for two different CAs all without AKIDs", func(t *testing.T) {
+		req := require.New(t)
+
+		root1 := newRootCa()
+		leaf1 := root1.NewLeafWithoutAKID()
+
+		root2 := newRootCa()
+		intermediate2 := root2.NewIntermediateWithoutAKID()
+		leaf2 := intermediate2.NewLeafWithoutAKID()
+
+		ret, err := AssembleServerChains([]*x509.Certificate{root1.cert, leaf1.cert, root2.cert, intermediate2.cert, leaf2.cert})
+
+		req.NoError(err)
+		req.NotNil(ret)
+		req.Len(ret, 2)
+
+		// root1 -> leaf2
+		req.Len(ret[0], 2)
+		req.Equal(ret[0][0], leaf1.cert)
+		req.Equal(ret[0][1], root1.cert)
+
+		// root2 -> intermediate2 -> leaf 2
+
+		req.Len(ret[1], 3)
+		req.Equal(ret[1][0], leaf2.cert)
+		req.Equal(ret[1][1], intermediate2.cert)
+		req.Equal(ret[1][2], root2.cert)
+	})
+
+	t.Run("returns 1 chain for ca>intermediate>leaf + random intermediates and CAs with AKIDs", func(t *testing.T) {
+		req := require.New(t)
+
+		root1 := newRootCa()
+		intermediate1 := root1.NewIntermediateWithAKID()
+
+		root2 := newRootCa()
+		intermediate2 := root2.NewIntermediateWithAKID()
+		leaf2 := intermediate2.NewLeafWithAKID()
+
+		root3 := newRootCa()
+		intermediate3 := root3.NewIntermediateWithAKID()
+
+		ret, err := AssembleServerChains([]*x509.Certificate{root1.cert, intermediate1.cert, root2.cert, intermediate2.cert, leaf2.cert, intermediate3.cert})
+
+		req.NoError(err)
+		req.NotNil(ret)
+		req.Len(ret, 1)
+
+		// root1 -> leaf2
+		req.Len(ret[0], 3)
+		req.Equal(ret[0][0], leaf2.cert)
+		req.Equal(ret[0][1], intermediate2.cert)
+		req.Equal(ret[0][2], root2.cert)
+	})
+
+	t.Run("returns 1 chain ca>intermediate>leaf with mixed AKID/no AKID", func(t *testing.T) {
```

Comment:
did i miss the test when the leaf doesn't match anything in the supplied chain? shouldn't there be a test that it returns nil? any other 'negative' type tests to add?

Reply:
Sure can add one.